src/DI/Factory.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace ILIAS\DI;
+
+use ILIAS\DI\Exceptions\InvalidClassException;
+use ILIAS\DI\Exceptions\NoSuchServiceException;
+
+/**
+ * Class Factory
+ * @package ILIAS\DI
+ *
+ * Create instances of classes using type hinting and the dependency injection container.
+ *
+ * @author Oskar Truffer <ot@studer-raimann.ch>
+ */
+class Factory {
```

Comment:
> Folklore says, ReflectionClass is slow. This is not too relevant for the background tasks, but might be a problem when this factory is used in ordinary requests. We should measure the impact of ReflectionClass before we make this generally available.

Injection needs reflection and the benchmarks I have found (admittedly they do not seem very thorough) indicate that the overhead generated by the reflection is neglectable. If we don't want reflection, we need to rename the "Dependency Injection" service to "Dependency Container".

> More severe: This pattern confuses "implements XYZ" with "is named XYZ". This is not the same. By announcing a dependency in a constructor, one (optimally) uses interfaces to describe the services an object depends on. This is the "implement XYZ" part. The array-access on the DIC, on the contrary, gives names to certain services. These are optimally only temporary/local when DI is used properly, but ILIAS is not there yet. The problem with this confusion is: the interfaces a service implements only accidentally coincide with its name. There already are services in ILIAS where this is in fact wrong ($tpl vs. ilTemplate). When announcing a dependency, I should only say: "I need some service implementing interface XYZ", but with this pattern I, in fact, say "I need the service named XYZ". This will prevent some moves in the future, e.g. having two services implementing the same interface.

Here we have the problem that I provided the interface with the implementation. The Interface provided in this PR stayed the same the implementation was changed in my implementation branch. This is a mistake on my part, sorry. The current implementation uses a dependency map to match interfaces to instantiations in the DIC regarding the class that requests the dependency. Thus it will be possible to have two implementations of an interface and hand them out depending on who requests the dependency (the redis + mysql usecase we discussed). 

This said I'm not in favor nor against keeping this interface in the DI and can move it to the BT as well. I thought it might be of interest to provide a basic injector to actually justify the services name as a first step. If you are still against keeping this and we get no other feedback by tomorrow I'll move it to the BT service.

Thanks a lot for the review! :-)